runtime/src/accounts_db.rs

```diff
@@ -4799,6 +4812,8 @@ impl AccountsDb {
                     store.slot(), *slot
                 );
                 let count = store.remove_account(account_info.stored_size, reset_accounts);
+                self.dirty_stores
+                    .insert((*slot, store.append_vec_id()), store.clone());
```

Comment:
hmmm it's not immediately obvious to me why removing an account from a storage entry should add the entire store to the `dirty_stores` list.  Maybe a comment would help?

I think the reasoning here is if for example:
1) the storage entry for slot `S` contains some pubkey `P`
2) Pubkey `P` has zero lamports in some slot `S' > S`.

 Then when the store for slot `S` hits `count ==0`, then this could potentially unblock the clean of the zero-lamport `P` in `S'`?

In this case, should we only insert into the dirty stores if the `count == 0` below?

Reply:
Well it's not just count==0, a shrink which removes an entry could also unblock it. Maybe the logic should be then to add if count==0 and then in shrink add the unref'ed key set there as well. I don't *think* there are any other cases.

Reply:
What if you have a 0-lamport account which is the newest entry for that pubkey in slot `S` for `X`. You also have a non-zero lamport update there for `Y`. Then in slot `S+1` you update `Y` to 0. Then the clean will deref `Y` from `S` and the 0-lamport update for `X` will keep the count=1, although now this store can be removed as long as `X` reference count in storage is 1 or if any other existing updates for it can be also removed.

Reply:
Ok. I think that's right. I changed to only inserting when count==0 such that the store is removed. That will free up those keys potentially in newer slots.

Reply:
> What if you have a 0-lamport account which is the newest entry for that pubkey in slot `S` for `X`. You also have a non-zero lamport update there for `Y`. Then in slot `S+1` you update `Y` to 0. Then the clean will deref `Y` from `S` and the 0-lamport update for `X` will keep the count=1, although now this store can be removed as long as `X` reference count in storage is 1 or if any other existing updates for it can be also removed.

Hmmm, is this above situation resolved?

So we had this situation:
```
<Slot S>                   <Slot S+1>         
X=0, Y=1                        Y=0
```

1) Slot `S` gets rooted, incurs a clean on `X` and `Y`. Account `X` cannot be removed because the `count` of slot `S`  starts at `2`, gets decremented by 1 for the zero lamport account `X`: https://github.com/solana-labs/solana/blob/43b09c4a24979b0d13f89002534797f815c56891/runtime/src/accounts_db.rs#L1834, but is ultimately not not zero in `calc_delete_dependencies`: https://github.com/solana-labs/solana/blob/43b09c4a24979b0d13f89002534797f815c56891/runtime/src/accounts_db.rs#L1506

2)`S+1` gets rooted and added to the `dirty_stores`, which incurs a clean on account `Y`. dropping the count of slot `S` to 1.  However, `Y` won't be deleted because its ref count is still 2 (due to the update in slot `S`), the ref count is only decremented when either slot `S` is dead or shrunk.

3) In shrink, there's two issues:<\br>
a) We may not shrink `S` because it doesn't meet the page size shrink criteria, even though not performing this shrink prevents the ref count of zero lamport slots from hitting zero, and thus is blocking a lot of zero lamport cleans in later slots.<\br>
b) We shrink slot `S`, but the store still has `count == 1` (the update to X is the latest update to 0), so we don't hit this case: https://github.com/solana-labs/solana/blob/43b09c4a24979b0d13f89002534797f815c56891/runtime/src/accounts_db.rs#L2205-L2207 and we *still* don't add `X` to the clean list.

Reply:
I wonder if the fix here is to say if in this check here: https://github.com/solana-labs/solana/blob/43b09c4a24979b0d13f89002534797f815c56891/runtime/src/accounts_db.rs#L1483, if `if account_infos.len() as u64 == *ref_count_from_storage == 1`, then we don't need to make the additional check in this else block: https://github.com/solana-labs/solana/blob/43b09c4a24979b0d13f89002534797f815c56891/runtime/src/accounts_db.rs#L1496-L1510 since we know there's no other slots that we are shielding

A test might be good to illustrate this situation here. I think there are similar ones, but the key difference here is we're cleaning after setting a root on S, and then cleaning after setting a root on S+1, instead of cleaning only once after rooting both S and S + 1

Reply:
> Doesn't this test do that?

Yeah, this test tests exactly this case, thanks!

>I don't think this is the case. When we shrink, the old store(s) should always have count==0 because all live accounts would be updated in it.

Ah yeah, I misread that bit of logic and thought it would only execute if the newly shrunk store had `count==0`, not the old unshrunken store, which will always have `count == 0` . So this will essentially add all the accounts in the old store for that slot to the candidate clean set.

Is it true that if we shrink unshrunken storage entry `U` to a shrunken storage `S`, we only need to add the zero-lamport pubkeys in `S` to the candidate clean set if the *only* remaining alive accounts in `S` are zero lamport themselves?

i.e. if we declared a 
```
let mut only_zero_lamport_accounts_alive = true;
```

Had a check here: 
https://github.com/solana-labs/solana/blob/43b09c4a24979b0d13f89002534797f815c56891/runtime/src/accounts_db.rs#L2107

```
if stored_account.account.account_meta.lamports != 0 {
     only_zero_lamport_accounts_alive = false;
}
```

And then add to the dirty stores here: https://github.com/solana-labs/solana/blob/43b09c4a24979b0d13f89002534797f815c56891/runtime/src/accounts_db.rs#L2205
```
// Add only the zero lamport accounts to be cleaned
if only_zero_lamport_accounts_alive && store.count() > 0 {
         self.dirty_stores
                            .insert((slot, store.append_vec_id()), store.clone());
}
```

Reasoning being:
1) If the post-shrunk store `S` has any non-zero lamport keys `K`, then when `K` is updated in later slots and `S` becomes dead, these keys will be added to the clean list (so don't need to add them in shrink).
2) If the post-shrunk store `S` *only* has zero lamport keys, then we can run into the edge case so we should add them to clean.

Reply:
Yea. This could be an optimization on the reducing the clean pubkey set, right? I think that could be a follow-up change. This already has a huge impact on the clean time, and a few extra keys are not that expensive to deal with.